### PR TITLE
fix(18574): fix new BigNumber() not a number: undefined for setApprovalForAll method

### PR DESCRIPTION
**Pull request by DDDDDanica**
Fixes:
https://github.com/MetaMask/metamask-extension/issues/18574

Error example:
https://metamask.sentry.io/share/issue/1082e57247a14a8b99fdc3347192ab51/

## Explanation
This error could possibly because when providing an ERC20 token with no decimals, we were trying to retrieve `tokenAmount`  but not able to finish `calcTokenAmount`.
However, we were not able to reproduce this error locally. 
What we will do is to push this fix to next release, and monitor sentry with same error 🙏 

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

### CODE REVIEW

File: ui/helpers/utils/token-util.js [hunk 1]
```diff
@@ -261,17 +261,15 @@ export async function getAssetDetails(
...
-    tokenDetails?.decimals &&
-    calcTokenAmount(
-      getTokenValueParam(tokenData),
-      tokenDetails?.decimals,
-    ).toString(10);
-
-  const decimals =
-    tokenDetails?.decimals && Number(tokenDetails.decimals?.toString(10));
+    tokenValue &&
+    tokenDecimals &&
```
Comment (Gudahtt) [inline ui/helpers/utils/token-util.js:269]: How does this handle tokens with `decimals` set to `0`?

Comment (DDDDDanica) [reply ui/helpers/utils/token-util.js:269]: As far as i can see, 0 decimals is handled by the usage of tokenDecimals.
- inside `calcTokenAmount`, decimal 0 is accepted 
```
  const multiplier = Math.pow(10, Number(decimals || 0));
```
- inside the decimals variable, it doesn't seem like 0 would impact the final result 
```
  const decimals = tokenDecimals && Number(tokenDecimals?.toString(10));
```
do you have any other concerns i didn't catch?

Comment (Gudahtt) [reply ui/helpers/utils/token-util.js:269]: Ah I see, this is a string right? I was thinking that if this was a number, 0 would be falsy and `calcTokenAmount` would never be run. But if it's a string, this should work fine.

**Review by Gudahtt — APPROVED** [review]
LGTM!

I am curious how users are getting in this situation though. I thought we didn't allow the decimals field to be unset? Maybe we should add metrics to detect this scenario if we're unsure what's causing it (in a separate PR I mean). And if we know of a way to add a token without setting a `decimals` field, we should fix that, and come up with a strategy for handling pre-existing token entries with that problem.

**Comment by DDDDDanica** [comment]
@Gudahtt thanks for reviewing. We were trying to reproduce the bug locally and the only trace we can get is when tokenDetails does not have a decimals passing in and we fall into undefined big number calc. 
@danjm were suggesting that it could be possible that user imported an ERC20 token with ‘setAppovalForAll’ method in an older version of MM where symbol and decimals are manually set. And when they interacting dapp with new version of MM we have, and initialize a setApprovalForAll action, it hits the undefined and throws error out. 
What I would do is 
1) monitor sentry with errors 2 weeks after release daily to see if any same error happens 
2) trace back which old version of MM we had that can manually add symbol and decimals to reproduce maybe.

**Review by danjm — APPROVED** [review]

**Comment by danjm** [comment]
> I am curious how users are getting in this situation though. I thought we didn't allow the decimals field to be unset?

The problem is moreso that there are tokens that have `decimals` but don't have a `value` parsable by `getTokenValueParam`

Danica mentioned one way that could happen. Another is that a user adds an NFT, and its contract mostly meets the NFT spec but it also has a `decimals` property.

**Pull request merged**